BAL-3498: enhance rule-engine evaluation logic with path comparison

shanegrouber opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Improved rule processing to deliver more robust value extraction and advanced comparison handling.
  - Added a new property for indicating path-based comparisons in specific rule operators.
  - Enhanced rule engine tests with support for path comparisons.
- **Refactor**
  - Enhanced internal rule validation logic for clearer, more consistent evaluation.
- **Chores**
  - Updated internal migration settings to support ongoing system maintenance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shanegrouber committed:
feat(rule-engine): enhance rule evaluation logic with path comparison

- Introduce isPathComparison flag in Rule schema for clearer distinction
- Update extractValue method to handle operators with and without path comparison
- Adjust validation logic to accommodate new extracted value structure

shanegrouber committed:
fix(rule-engine): correct path comparison validation logic

- Ensure 'isPathComparison' is only true when present in the rule
- Update condition to prevent false positives in comparison checks

tomer-shvadron commented on packages/common/src/rule-engine/operators/helpers.ts:
```diff
@@ -39,12 +39,36 @@ export abstract class BaseOperator<TDataValue = Primitive, TConditionValue = Pri
 
   extractValue(data: unknown, rule: Rule) {
     const value = get(data, rule.key);
+    const OPERATORS_WITHOUT_PATH_COMPARISON = [
+      OPERATION.AML_CHECK,
+      OPERATION.BETWEEN,
+      OPERATION.LAST_YEAR,
+    ] as const;
```
There's no need to re-declare this array every time the `extractValue` function is triggered. Extract it outside the class

tomer-shvadron commented on services/workflows-service/src/rule-engine/core/rule-engine.ts:
```diff
@@ -23,11 +23,19 @@ export const validateRule = (rule: Rule, data: any): RuleResult => {
+  const extractedValue = operator.extractValue(data, rule);
+
+  const isPathComparison =
+    typeof extractedValue === 'object' &&
```
use the `isObject` function from `@ballerine/common` instead, see if you can use the `?.` notation after that

shanegrouber committed:
refactor(rule-engine): improve operator extraction logic and constants usage

- Move OPERATORS_WITHOUT_PATH_COMPARISON to constants for better reusability
- Refactor extraction logic to use isObject utility for clarity
- Update imports to reflect the new constants structure

shanegrouber committed:
Merge branch 'dev' into BAL-3498

shanegrouber committed:
fix(rule-engine): improve rule extraction validation

- Refactor validation logic for extracted values
- Utilize 'in' operator for better readability and accuracy

shanegrouber committed:
feat(tests): enhance rule engine tests with path comparison functionality

- Add isPathComparison flag to various rule definitions
- Implement new unit tests for path comparison scenarios
- Update integration tests to include path comparison in validation

shanegrouber committed:
Merge branch 'BAL-3498' of github.com:ballerine-io/ballerine into BAL-3498
